server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -145,69 +145,79 @@ public void start() {
         }
 
         protected void doStart(ClusterState clusterState) {
-            final Predicate<ClusterState> masterChangePredicate = MasterNodeChangePredicate.build(clusterState);
-            final DiscoveryNodes nodes = clusterState.nodes();
-            if (nodes.isLocalNodeElectedMaster() || localExecute(request)) {
-                // check for block, if blocked, retry, else, execute locally
-                final ClusterBlockException blockException = checkBlock(request, clusterState);
-                if (blockException != null) {
-                    if (!blockException.retryable()) {
-                        listener.onFailure(blockException);
-                    } else {
-                        logger.trace("can't execute due to a cluster block, retrying", blockException);
-                        retry(blockException, newState -> {
-                            ClusterBlockException newException = checkBlock(request, newState);
-                            return (newException == null || !newException.retryable());
-                        });
-                    }
-                } else {
-                    ActionListener<Response> delegate = new ActionListener<Response>() {
-                        @Override
-                        public void onResponse(Response response) {
-                            listener.onResponse(response);
+            try {
+                final Predicate<ClusterState> masterChangePredicate = MasterNodeChangePredicate.build(clusterState);
+                final DiscoveryNodes nodes = clusterState.nodes();
+                if (nodes.isLocalNodeElectedMaster() || localExecute(request)) {
+                    // check for block, if blocked, retry, else, execute locally
+                    final ClusterBlockException blockException = checkBlock(request, clusterState);
+                    if (blockException != null) {
+                        if (!blockException.retryable()) {
+                            listener.onFailure(blockException);
+                        } else {
+                            logger.trace("can't execute due to a cluster block, retrying", blockException);
+                            retry(blockException, newState -> {
+                                try {
+                                    ClusterBlockException newException = checkBlock(request, newState);
+                                    return (newException == null || !newException.retryable());
+                                } catch (Exception e) {
+                                    // accept state as block will be rechecked by doStart() and listener.onFailure() then called
+                                    logger.trace("exception occurred during cluster block checking, accepting state", e);
+                                    return true;
```

Comment:
While it might be tricky to see in this diff what's actually changed: The three lines above here have been added.